### ISSUE TITLE
False positive when reporting E0641 (it compiles correctly)

### ISSUE THREAD

**Issue by rodrigocfd** [rust-lang/rust-analyzer#18214]
I'm writing some code that interacts with Windows C code, something like this:

```rust
struct Foo {}

struct Wrap<'a> {
    foo: &'a mut Foo,
}

fn main() {
    let lparam: isize = 0; // this value comes from Windows C code

    let wrap = Wrap {
        foo: unsafe { &mut *(lparam as *mut _) }, // E0641
    };
}
```

Screenshot:

![Image](https://github.com/user-attachments/assets/2fa10853-ecd8-4372-ba1a-6d981678e392)

The error reported by Rust Analyzer is [E0641](https://doc.rust-lang.org/stable/error_codes/E0641.html). However, this code [compiles correctly](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bc44616fab5bd303a9a94b12c7ee9e40).

This error started showing this month (maybe)... so, it's probably some sort of regression.

Note: in the example, if I declare `lparam` as a pointer (instead of `isize`), no error is reported. However, in Windows events, `LPARAM` doesn't necessarily has a pointer, so `isize` is semantically more correct.

Tooling:

* rust-analyzer version: 0.3.2129-standalone (822644d97 2024-09-29)
* rustc 1.81.0 (eeb90cda1 2024-09-04)
* VSCode 1.93.1
